Config schema validation

pSpitzner opened the pull request:
Good to go imho.

Remaining tasks:
- [x] Support more convenient inbox name via yaml
- [x] eyconf attribute dict rework so that "import" config section handled correctly [#16](https://github.com/semohr/eyconf/pull/16)
- [x] what happens after config error?
      - container starts,
      - but shows a config error on first page visit.
      - after that I dont care. ideally, add info that a container restart is needed. 
- [x] docs:
    - screenshot of frontend config error
    - what to do. 
- [x] Create a small benchmark for the new config validation (how much overhead)
- [ ] ~~Restart workers after config change, for live reload of config~~ -> separate PR

pSpitzner committed:
Added eyconf dependency and drafted dataclasses for schema validation

pSpitzner committed:
fixed failing tests due to config validation. however, i think our mixins do not work
as we think - we get some ~ 30 tests failing without those two changes.
it seems modifications to the config object survive `refresh_config()` calls,
at least in the test suite!

pSpitzner committed:
started new beets config from scratch. did some type experiments that might better
end up in eyconf.

pSpitzner commented:
So, the last commit is an interesting one. @semohr  we might want to put most of this logic into eyconf, as the problem it tries to solve is very general:
- sometimes we have parts of configs for which we do not have a schema (here beets stuff)
- we still want those to be part of your dataclass, accessible as attributes via `.my_key`
- we want typing on all known keys and type errors on unknown ones
- and need utilities to go back and forth between the (both deeply nested) dataclass and the dict representations, in order to feed back into confuse.

Here is what we can do with the latest commit:


```python
from beets_flask.config.beets_config_schema import BeetsSchema
from beets_flask.config.beets_config_new import BeetsFlaskConfig
from beets_flask.config.type_shenanigans import (
    dict_to_dataclass_with_known_schema,
    mixed_dataclass_to_dict,
    __format_with_indentation,
)

config = BeetsFlaskConfig()

# conversion to dicts
data = mixed_dataclass_to_dict(config)

# manipulate the dict, which can then be converted back to dataclass
# when converting back, unknown attributes get added
# missing attributes get set to defaults
data["gui"]["library"]["readonly"] = True
data["gui"]["library"]["foo"] = "asdf"
data["gui"]["library"].pop("artist_separators")

# either create a copy
config_new = dict_to_dataclass_with_known_schema(schema=BeetsSchema, data=data)

# or apply in place, same as `config.update_from_dict(data)`
dict_to_dataclass_with_known_schema(
    schema=BeetsSchema,
    data=data,
    existing_instance=config,
)

# access via dots works and is typed
config.gui.inbox.folders["placeholder"].path
config_new.gui.inbox.folders["placeholder"].path


# readable output, dynamic attributes marked by `*`
print(config)

# dicts too, but not marked
print(__format_with_indentation(repr(mixed_dataclass_to_dict(config))))
```

Gives us
```python
BeetsFlaskConfig(
    gui=BeetsFlaskSchema(
        inbox=InboxSection(
            ignore='_use_beets_ignore',
            debounce_before_autotag=30,
            folders={
                'placeholder': InboxFolder(
                    name='Please check your config!',
                    path='/music/inbox',
                    auto_threshold=None,
                    autotag=False
                )
            }
        ),
        library=LibrarySection(
            readonly=True,
            artist_separators=[
                ',',
                ';',
                '&'
            ],
            *foo='asdf'
        ),
        terminal=TerminalSection(
            start_path='/repo'
        ),
        num_preview_workers=4
    )
)


{
    'gui':{
        'inbox':{
            'ignore': '_use_beets_ignore',
            'debounce_before_autotag': 30,
            'folders':{
                'placeholder':{
                    'name': 'Please check your config!',
                    'path': '/music/inbox',
                    'auto_threshold': None,
                    'autotag': False
                }
            }
        },
        'library':{
            'readonly': True,
            'artist_separators':[
                ',',
                ';',
                '&'
            ],
            'foo': 'asdf'
        },
        'terminal':{
            'start_path': '/repo'
        },
        'num_preview_workers': 4
    }
}
```

semohr committed:
Refactored config class to newer leaner format using eyconf.

Co-authored-by: pSpitzner <github@makeitso.one>

semohr committed:
get_config now uses the new format.

Co-authored-by: pSpitzner <github@makeitso.one>

pSpitzner committed:
- fixed the remaining tests
- renamed `refresh` to `reload` to indicate the io happening,
  and `to_confuse` to `beets_config` as property. This felt more consistent, as we
  cannot afford to recreate beets' config every time we access its config (too slow)

pSpitzner committed:
A bit of cleanup:
- light refactoring
- consistent str type for autotag config option
- started using new schema via py2ts in frontend

pSpitzner committed:
if we do not reset the beets config so often, we need to be more careful with
overlays

semohr commented on docs/configuration.md:
```diff
@@ -30,7 +30,7 @@ You may add multiple inboxes, each may have a different purpose.
 The `autotag` setting determines how the files in the inbox are processed by beets-flask.
 
-- <i data-lucide="inbox"></i> `"no"` you have to do everything manually.
+- <i data-lucide="inbox"></i> `"off"` you have to do everything manually.
```
We should create some helpers for deprecation warnings.

pSpitzner committed:
trying to fix file io tests

semohr committed:
Created a deprecation warning helper. Should help us identify
depredations easier.

semohr commented on frontend/src/routes/version.tsx:
```diff
@@ -119,7 +119,7 @@ function DataSources() {
             </Typography>
             <Box sx={{ display: "flex", flexWrap: "wrap", gap: 2 }}>
-                {config.data_sources.map((source) => (
+                {config.beets_meta_sources.map((source) => (
```
Im not too sure how I feel about this rename. Shouldn't it be clear we talk about beets? `data_sources` was descriptive enough for me and it mirrors how beets names these things.

pSpitzner commented on backend/beets_flask/server/routes/config.py:
```diff
@@ -16,46 +15,24 @@
-            "plugins": config["plugins"].as_str_seq(),
-            "data_sources": data_sources,
-            "beets_version": beets_version,
+            # utility getters, could become part of the schema
```
see below, just thinking out loud how to make clear whats (not) indcluded via the schema

pSpitzner committed:
added tests for config and included full validation in reload
(before we did half of the validation, schema only, that felt too inbetween)

pSpitzner committed:
Added changelog entry and configuration error if a configured inbox folder
does not exist

pSpitzner committed:
closes #235

pSpitzner committed:
updated eyconf, but more work upstream needed, fixed some tests, typed import section

pSpitzner committed:
removed workarounds since we updated eyconf

pSpitzner committed:
consistent naming `metadata_sources`

pSpitzner committed:
wip: further config tweaks

pSpitzner committed:
updated config cloading to match eyconf PR
[PR 16](https://github.com/semohr/eyconf/pull/16)
once we updated eyconf, this commit fixes our tests.

pSpitzner committed:
inbox folders can now use headings as name fallbacks

pSpitzner committed:
fixed a bug on inbox folder validation, but that required a bit of more test mixins.
one still failing.

pSpitzner committed:
cleaner config override in test suite,
bumped eyconf dependency

pSpitzner committed:
added benchmark and fixed commit_to_beets.
our config with validation is at most some 3ms slower (5ms to 8ms), so seems fine since
it should be a one-time call.

pSpitzner committed:
added docs for schema validation, and updated frontend types

pSpitzner committed:
fix "_use_heading" frontend type

pSpitzner committed:
linting, formatting and dependencies

pSpitzner committed:
updated changelog
